src/librustc_resolve/lib.rs

```diff
@@ -2700,20 +2702,51 @@ impl<'a> Resolver<'a> {
                         }
                         return (err, candidates);
                     },
-                    _ if ns == ValueNS && is_struct_like(def) => {
-                        if let Def::Struct(def_id) = def {
-                            if let Some((ctor_def, ctor_vis))
-                                    = this.struct_constructors.get(&def_id).cloned() {
-                                if is_expected(ctor_def) && !this.is_accessible(ctor_vis) {
-                                    err.span_label(span, format!("constructor is not visible \
-                                                                   here due to private fields"));
-                                }
+                    (Def::Struct(def_id), _) if ns == ValueNS => {
+                        if let Some((ctor_def, ctor_vis))
+                                = this.struct_constructors.get(&def_id).cloned() {
+                            let accessible_ctor = this.is_accessible(ctor_vis);
+                            if is_expected(ctor_def) && !accessible_ctor {
+                                err.span_label(span, format!("constructor is not visible \
+                                                              here due to private fields"));
+                            } else if accessible_ctor {
+                                let block = match ctor_def {
+                                    Def::StructCtor(_, CtorKind::Fn) => "(/* fields */)",
+                                    Def::StructCtor(_, CtorKind::Const) => "",
+                                    Def::Struct(..) => " { /* fields */ }",
+                                    def => bug!("found def `{:?}` when looking for a ctor", def),
+                                };
+                                err.span_label(span, format!("did you mean `{}{}`?",
+                                                             path_str,
+                                                             block));
                             }
+                        } else {
+                            err.span_label(span, format!("did you mean `{} {{ /* fields */ }}`?",
+                                                         path_str));
                         }
+                        return (err, candidates);
+                    }
+                    (Def::VariantCtor(_, CtorKind::Const), _) => {
+                        err.span_label(span, format!("did you mean `{}`?", path_str));
+                        return (err, candidates);
+                    }
+                    (Def::VariantCtor(_, CtorKind::Fn), _) => {
+                        err.span_label(span, format!("did you mean `{}( /* fields */ )`?",
+                                                      path_str));
+                        return (err, candidates);
+                    }
+                    (Def::VariantCtor(_, CtorKind::Fictive), _) => {
```

Comment:
These three `VariantCtor` arms look like they give tautological recommendations, I don't think they are ever useful (at least in realistic code).